verification/cbmc/sources/make_common_data_structures.c

```diff
@@ -137,12 +141,82 @@ void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list) {
 }
 
 void ensure_cryptosdk_edk_list_has_allocated_list_elements(struct aws_array_list *list) {
+    if (!aws_cryptosdk_edk_list_is_valid(list)) {
+        return;
+    }
     for (size_t i = 0; i < list->length; ++i) {
         struct aws_cryptosdk_edk *data = (struct aws_cryptosdk_edk *)list->data;
         ensure_cryptosdk_edk_has_allocated_members(&(data[i]));
     }
 }
 
+struct aws_cryptosdk_hdr *ensure_nondet_hdr_has_allocated_members() {
+    struct aws_cryptosdk_hdr *hdr = malloc(sizeof(*hdr));
+    if (hdr == NULL) {
+        return NULL;
+    }
+    hdr->alloc = nondet_bool() ? NULL : can_fail_allocator();
+    enum aws_cryptosdk_alg_id alg_id;
+    hdr->alg_id    = alg_id;
+    hdr->frame_len = malloc(sizeof(hdr->frame_len));
+
+    struct aws_byte_buf *iv = malloc(sizeof(*iv));
+    if (iv) {
+        hdr->iv = *iv;
+    }
+    struct aws_byte_buf *auth_tag = malloc(sizeof(*auth_tag));
+    if (auth_tag) {
+        hdr->auth_tag = *auth_tag;
+    }
+    struct aws_byte_buf *message_id = malloc(sizeof(*message_id));
+    if (message_id) {
+        hdr->message_id = *message_id;
+    }
+    struct aws_byte_buf *alg_suite_data = malloc(sizeof(*alg_suite_data));
+    if (alg_suite_data) {
+        hdr->alg_suite_data = *alg_suite_data;
+    }
+    struct aws_hash_table *enc_ctx = malloc(sizeof(*enc_ctx));
+    if (enc_ctx) {
+        hdr->enc_ctx = *enc_ctx;
+    }
+    struct aws_array_list *edk_list = malloc(sizeof(*edk_list));
+    if (edk_list) {
+        hdr->edk_list = *edk_list;
+    }
+
+    hdr->auth_len = malloc(sizeof(hdr->auth_len));
```

Comment:
I'd recommend
```c
if (hdr != NULL) {
    hdr->alloc = nondet_bool() ? NULL : can_fail_allocator();
    hdr->frame_len = malloc(sizeof(hdr->frame_len));
    ensure_byte_buf_has_allocated_buffer_member(&hdr->iv);
    ensure_byte_buf_has_allocated_buffer_member(&hdr->auth_tag);
    ensure_byte_buf_has_allocated_buffer_member(&hdr->message_id);
    ensure_byte_buf_has_allocated_buffer_member(&hdr->alg_suite_data);
    ensure_allocated_hash_table(&hdr->enc_ctx, /* YOU NEED A BOUND HERE. */);
    ensure_array_list_has_allocated_data_member(&hdr->edk_list);
}
```
If a structure field is not a pointer, you don't need to allocate or make sure is non-deterministic, CBMC will do that automatically. You need to make sure the pointer structures are properly allocated, and I'd suggest using the existing ensure functions.

Reply:
Oh I see! Updated as suggested!